components/auth-dialog.tsx

```diff
@@ -1,210 +1,169 @@
-                            onChange={(e) =>
-                              setConfirmPassword(e.target.value.trim())
-                            }
-                          />
-                        </div>
-                        <div className="pt-7">
-                          <Button className="w-full" onClick={signUp}>
-                            Next{" "}
-                            <ArrowRight
-                              className="w-4 h-4 ml-2"
-                              strokeWidth={3}
-                            />
-                          </Button>
-                        </div>
-                      </div>
-                    </div>
-                  );
-                case "success":
-                  return (
-                    <div className="block w-full px-28 text-center">
-                      "Thanks for signing up! Please check your email for a
-                      verification link."
-                    </div>
-                  );
-              }
-            })()}
-            {error && <div className="text-red-500 text-sm mb-4">{error}</div>}
-          </div>
-        </DialogContent>
-      </Dialog>
-    );
-  }
\ No newline at end of file
+	const [step, setStep] = useState<"email" | "log-in" | "sign-up" | "success">(
+		"email",
+	);
+	const [error, setError] = useState("");
+	const [email, setEmail] = useState("");
+	const [password, setPassword] = useState("");
+	const [confirmPassword, setConfirmPassword] = useState("");
+	const [isOpen, setIsOpen] = useState(false);
+	const [username, setUsername] = useState("");
+
+	useEffect(() => {
+		if (!isOpen) {
+			setStep("email");
+			setError("");
+			setEmail("");
+			setPassword("");
+			setConfirmPassword("");
+			setUsername("");
+		}
+	}, [isOpen]);
+
+	const submitEmail = async () => {
+		setError("");
+
+		const emailRegex = /^[^\s@]+@[^\s@]+\.[^\s@]+$/;
+		if (!emailRegex.test(email)) {
+			setError("Please enter a valid email address");
+			setStep("email");
+			return;
+		}
+
+		try {
+			const userExists = await checkUserAction(email);
+			setStep(userExists ? "log-in" : "sign-up");
+		} catch (error) {
+			console.error("Error checking user existence:", error);
+			setError("Failed to check user existence. Please try again.");
+			setStep("email");
+		}
+	};
+
+	const signIn = async () => {
+		setError("");
+		const formData = new FormData();
+		formData.append("email", email);
+		formData.append("password", password);
+
+		const result = await signInAction(formData);
+
+		if (!result.success) {
+			setError(result.message);
+		} else {
+			setIsOpen(false);
+		}
+	};
+
+	const signUp = async () => {
+		setError("");
+		if (password !== confirmPassword) {
+			setError("Passwords do not match");
+			return;
+		}
+
+		if (password.length < 6) {
+			setError("Password should be at least 6 characters long");
+			return;
+		}
+
+		//TODO: check if username is already taken
+		if (username.length < 10) {
```

Comment:
![](https://media1.tenor.com/m/KAY6mFj4vcYAAAAd/dhsrp-tape.gif)
```suggestion
		if (username.length < 5) {
```